// uniforms-bridge-json-schema/es5/JSONSchemaBridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var invariant_1 = tslib_1.__importDefault(require("invariant"));
var cloneDeep_1 = tslib_1.__importDefault(require("lodash/cloneDeep"));
var get_1 = tslib_1.__importDefault(require("lodash/get"));
var lowerCase_1 = tslib_1.__importDefault(require("lodash/lowerCase"));
var memoize_1 = tslib_1.__importDefault(require("lodash/memoize"));
var omit_1 = tslib_1.__importDefault(require("lodash/omit"));
var upperFirst_1 = tslib_1.__importDefault(require("lodash/upperFirst"));
var uniforms_1 = require("uniforms");
function resolveRef(reference, schema) {
    invariant_1.default(reference.startsWith('#'), 'Reference is not an internal reference, and only such are allowed: "%s"', reference);
    var resolvedReference = reference
        .split('/')
        .filter(function (part) { return part && part !== '#'; })
        .reduce(function (definition, next) { return definition[next]; }, schema);
    invariant_1.default(resolvedReference, 'Reference not found in schema: "%s"', reference);
    return resolvedReference;
}
var propMapper = {
    maxItems: 'maxCount',
    minItems: 'minCount',
    maximum: 'max',
    minimum: 'min',
    multipleOf: 'step',
};
function distinctSchema(schema) {
    if (schema.type === 'object') {
        return schema;
    }
    if (schema.$ref) {
        return tslib_1.__assign(tslib_1.__assign({}, schema), resolveRef(schema.$ref, schema));
    }
    return schema;
}
function extractValue() {
    var xs = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        xs[_i] = arguments[_i];
    }
    return xs.reduce(function (x, y) {
        return x === false || x === null ? '' : x !== true && x !== undefined ? x : y;
    });
}
function pathToName(path) {
    path = path.startsWith('/')
        ? path.replace(/\//g, '.').replace(/~0/g, '~').replace(/~1/g, '/')
        : path
            .replace(/\[('|")(.+?)\1\]/g, '.$2')
            .replace(/\[(.+?)\]/g, '.$1')
            .replace(/\\'/g, "'");
    return path.slice(1);
}
function toHumanLabel(label) {
    return upperFirst_1.default(lowerCase_1.default(label));
}
var JSONSchemaBridge = /** @class */ (function (_super) {
    tslib_1.__extends(JSONSchemaBridge, _super);
    function JSONSchemaBridge(schema, validator) {
        var _this = _super.call(this) || this;
        _this.schema = schema;
        _this.validator = validator;
        _this._compiledSchema = {};
        _this.schema = distinctSchema(schema);
        _this._compiledSchema[''] = _this.schema;
        // Memoize for performance and referential equality.
        _this.getField = memoize_1.default(_this.getField.bind(_this));
        _this.getSubfields = memoize_1.default(_this.getSubfields.bind(_this));
        _this.getType = memoize_1.default(_this.getType.bind(_this));
        return _this;
    }
    JSONSchemaBridge.prototype.getError = function (name, error) {
        var _a, _b;
        var nameParts = uniforms_1.joinName(null, name);
        var rootName = uniforms_1.joinName(nameParts.slice(0, -1));
        var baseName = nameParts[nameParts.length - 1];
        return (
        // FIXME: Correct type for `error`.
        ((_b = (_a = error === null || error === void 0 ? void 0 : error.details) === null || _a === void 0 ? void 0 : _a.find) === null || _b === void 0 ? void 0 : _b.call(_a, function (detail) {
            var _a;
            var path = pathToName((_a = detail.instancePath) !== null && _a !== void 0 ? _a : detail.dataPath);
            return (name === path ||
                (rootName === path && baseName === detail.params.missingProperty));
        })) || null);
    };
    JSONSchemaBridge.prototype.getErrorMessage = function (name, error) {
        var scopedError = this.getError(name, error);
        return (scopedError === null || scopedError === void 0 ? void 0 : scopedError.message) || '';
    };
    JSONSchemaBridge.prototype.getErrorMessages = function (error) {
        if (!error) {
            return [];
        }
        var details = error.details;
        return Array.isArray(details)
            ? details.map(function (error) { return error.message; })
            : [error.message || error];
    };
    JSONSchemaBridge.prototype.getField = function (name) {
        var _this = this;
        return uniforms_1.joinName(null, name).reduce(function (definition, next, nextIndex, array) {
            var previous = uniforms_1.joinName(array.slice(0, nextIndex));
            var isRequired = get_1.default(definition, 'required', get_1.default(_this._compiledSchema, [previous, 'required'], [])).includes(next);
            var _key = uniforms_1.joinName(previous, next);
            var _definition = _this._compiledSchema[_key] || {};
            if (next === '$' || next === '' + parseInt(next, 10)) {
                invariant_1.default(definition.type === 'array', 'Field not found in schema: "%s"', name);
                definition = Array.isArray(definition.items)
                    ? definition.items[parseInt(next, 10)]
                    : definition.items;
            }
            else if (definition.type === 'object') {
                invariant_1.default(definition.properties, 'Field properties not found in schema: "%s"', name);
                definition = definition.properties[next];
            }
            else {
                var _a = [
                    'allOf',
                    'anyOf',
                    'oneOf',
                ]
                    .filter(function (key) { return definition[key]; })
                    .map(function (key) {
                    // FIXME: Correct type for `definition`.
                    var localDef = definition[key].map(function (subSchema) {
                        return subSchema.$ref
                            ? resolveRef(subSchema.$ref, _this.schema)
                            : subSchema;
                    });
                    return localDef.find(function (_a) {
                        var _b = _a.properties, properties = _b === void 0 ? {} : _b;
                        return properties[next];
                    });
                })[0], _b = _a === void 0 ? {} : _a, _c = _b.properties, combinedDefinition = _c === void 0 ? {} : _c;
                definition = combinedDefinition[next];
            }
            invariant_1.default(definition, 'Field not found in schema: "%s"', name);
            if (definition.$ref) {
                definition = resolveRef(definition.$ref, _this.schema);
            }
            ['allOf', 'anyOf', 'oneOf'].forEach(function (key) {
                if (definition[key]) {
                    // FIXME: Correct type for `definition`.
                    _definition[key] = definition[key].map(function (def) {
                        return def.$ref ? resolveRef(def.$ref, _this.schema) : def;
                    });
                }
            });
            // Naive computation of combined type, properties and required
            var combinedPartials = []
                .concat(_definition.allOf, _definition.anyOf, _definition.oneOf)
                .filter(Boolean);
            if (combinedPartials.length) {
                var localProperties_1 = definition.properties
                    ? tslib_1.__assign({}, definition.properties) : {};
                var localRequired_1 = definition.required
                    ? definition.required.slice()
                    : [];
                combinedPartials.forEach(function (_a) {
                    var properties = _a.properties, required = _a.required, type = _a.type;
                    if (properties) {
                        Object.assign(localProperties_1, properties);
                    }
                    if (required) {
                        localRequired_1.push.apply(localRequired_1, required);
                    }
                    if (type && !_definition.type) {
                        _definition.type = type;
                    }
                });
                if (Object.keys(localProperties_1).length > 0) {
                    _definition.properties = localProperties_1;
                }
                if (localRequired_1.length > 0) {
                    _definition.required = localRequired_1;
                }
            }
            _this._compiledSchema[_key] = Object.assign(_definition, { isRequired: isRequired });
            return definition;
        }, this.schema);
    };
    JSONSchemaBridge.prototype.getInitialValue = function (name, props) {
        if (props === void 0) { props = {}; }
        var _a = this.getField(name), _default = _a.default, _type = _a.type;
        var _b = this._compiledSchema[name], _c = _b.default, defaultValue = _c === void 0 ? _default !== undefined
            ? _default
            : get_1.default(this.schema.default, name) : _c, _d = _b.type, type = _d === void 0 ? _type : _d;
        if (defaultValue !== undefined) {
            return cloneDeep_1.default(defaultValue);
        }
        if (type === 'array') {
            var item = this.getInitialValue(uniforms_1.joinName(name, '0'));
            var items = props.initialCount || 0;
            return Array(items).fill(item);
        }
        if (type === 'object') {
            return {};
        }
        return undefined;
    };
    JSONSchemaBridge.prototype.getProps = function (name, props) {
        if (props === void 0) { props = {}; }
        var _a = this.getField(name), uniforms = _a.uniforms, field = tslib_1.__rest(_a, ["uniforms"]);
        var _b = omit_1.default(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({}, field), uniforms), this._compiledSchema[name]), ['default', 'format', 'type']), enum_ = _b.enum, isRequired = _b.isRequired, title = _b.title, ready = tslib_1.__rest(_b, ["enum", "isRequired", "title"]);
        if (enum_) {
            ready.allowedValues = enum_;
        }
        if (field.type === 'number') {
            ready.decimal = true;
        }
        if (uniforms && uniforms.type !== undefined) {
            ready.type = uniforms.type;
        }
        if (ready.required === undefined) {
            ready.required = isRequired;
        }
        ready.label = extractValue(ready.label, title, toHumanLabel(uniforms_1.joinName(null, name).slice(-1)[0]));
        var options = props.options || ready.options;
        if (options) {
            if (!Array.isArray(options)) {
                ready.transform = function (value) { return options[value]; };
                ready.allowedValues = Object.keys(options);
            }
            else {
                ready.transform = function (value) {
                    return options.find(function (option) { return option.value === value; }).label;
                };
                ready.allowedValues = options.map(function (option) { return option.value; });
            }
        }
        Object.keys(ready).forEach(function (key) {
            if (key in propMapper) {
                var newKey = propMapper[key];
                ready[newKey] = ready[key];
                delete ready[key];
            }
        });
        return ready;
    };
    JSONSchemaBridge.prototype.getSubfields = function (name) {
        if (name === void 0) { name = ''; }
        var field = this.getField(name);
        var _a = this._compiledSchema[name], _b = _a.properties, properties = _b === void 0 ? field.properties : _b, _c = _a.type, type = _c === void 0 ? field.type : _c;
        if (type === 'object' && properties) {
            return Object.keys(properties);
        }
        return [];
    };
    JSONSchemaBridge.prototype.getType = function (name) {
        var _a = this.getField(name), _type = _a.type, fieldFormat = _a.format;
        var _b = this._compiledSchema[name].type, fieldType = _b === void 0 ? _type : _b;
        if (fieldFormat === 'date-time') {
            return Date;
        }
        if (fieldType === 'string') {
            return String;
        }
        if (fieldType === 'number') {
            return Number;
        }
        if (fieldType === 'integer') {
            return Number;
        }
        if (fieldType === 'object') {
            return Object;
        }
        if (fieldType === 'array') {
            return Array;
        }
        if (fieldType === 'boolean') {
            return Boolean;
        }
        invariant_1.default(fieldType !== 'null', 'Field "%s" can not be represented as a type null', name);
        return fieldType;
    };
    JSONSchemaBridge.prototype.getValidator = function () {
        return this.validator;
    };
    return JSONSchemaBridge;
}(uniforms_1.Bridge));
exports.default = JSONSchemaBridge;
