// uniforms-antd/es5/ErrorsField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var uniforms_1 = require("uniforms");
var defaultStyle = {
    backgroundColor: 'rgba(255, 85, 0, 0.2)',
    border: '1px solid rgb(255, 85, 0)',
    borderRadius: '2px',
    margin: '20px 0px',
    padding: '10px',
};
var rowStyle = { margin: '3px' };
function ErrorsField(_a) {
    var children = _a.children, _b = _a.style, style = _b === void 0 ? defaultStyle : _b, props = tslib_1.__rest(_a, ["children", "style"]);
    var _c = uniforms_1.useForm(), error = _c.error, schema = _c.schema;
    return !error && !children ? null : (react_1.default.createElement("div", tslib_1.__assign({ style: style }, uniforms_1.filterDOMProps(props)),
        children,
        react_1.default.createElement("ul", null, schema.getErrorMessages(error).map(function (message, index) { return (react_1.default.createElement("li", { key: index, style: rowStyle }, message)); }))));
}
exports.default = ErrorsField;
