// uniforms-antd/es5/ListItemField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var uniforms_1 = require("uniforms");
var AutoField_1 = tslib_1.__importDefault(require("./AutoField"));
var ListDelField_1 = tslib_1.__importDefault(require("./ListDelField"));
var delStyle = {
    float: 'right',
    marginBottom: '10px',
    marginLeft: '10px',
    marginRight: '6px',
    width: '20px',
};
var itemStyle = { marginBottom: '24px', overflow: 'hidden' };
var dividerStyle = {
    borderBottom: '1px solid #DDD',
    height: '20px',
    marginTop: '-8px',
};
var childrenStyle = { width: '100%' };
function ListItem(_a) {
    var _b = _a.children, children = _b === void 0 ? react_1.default.createElement(AutoField_1.default, { label: null, name: "" }) : _b;
    return (react_1.default.createElement("div", null,
        react_1.default.createElement("div", { style: delStyle },
            react_1.default.createElement(ListDelField_1.default, { className: "top aligned", name: "" })),
        react_1.default.createElement("div", { style: itemStyle },
            react_1.default.createElement("div", { style: dividerStyle })),
        react_1.default.createElement("div", { style: childrenStyle }, children)));
}
exports.default = uniforms_1.connectField(ListItem);
