// @datapunt/matomo-tracker-react/lib/useMatomo.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = require("react");
var MatomoContext_1 = __importDefault(require("./MatomoContext"));
var useOutboundClickListener_1 = __importDefault(require("./utils/useOutboundClickListener"));
function useMatomo() {
    var instance = react_1.useContext(MatomoContext_1.default);
    var trackPageView = react_1.useCallback(function (params) { return instance === null || instance === void 0 ? void 0 : instance.trackPageView(params); }, [instance]);
    var trackEvent = react_1.useCallback(function (params) { return instance === null || instance === void 0 ? void 0 : instance.trackEvent(params); }, [instance]);
    var trackEvents = react_1.useCallback(function () { return instance === null || instance === void 0 ? void 0 : instance.trackEvents(); }, [instance]);
    var trackSiteSearch = react_1.useCallback(function (params) { return instance === null || instance === void 0 ? void 0 : instance.trackSiteSearch(params); }, [instance]);
    var trackLink = react_1.useCallback(function (params) { return instance === null || instance === void 0 ? void 0 : instance.trackLink(params); }, [instance]);
    var enableLinkTracking = react_1.useCallback(function () {
        if (instance) {
            useOutboundClickListener_1.default(instance);
        }
    }, [instance]);
    var pushInstruction = react_1.useCallback(function (name) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        instance === null || instance === void 0 ? void 0 : instance.pushInstruction.apply(instance, __spreadArrays([name], args));
    }, [instance]);
    return {
        trackEvent: trackEvent,
        trackEvents: trackEvents,
        trackPageView: trackPageView,
        trackSiteSearch: trackSiteSearch,
        trackLink: trackLink,
        enableLinkTracking: enableLinkTracking,
        pushInstruction: pushInstruction,
    };
}
exports.default = useMatomo;
